fastly/base_fastly_service_v1.go

```diff
@@ -204,54 +217,54 @@ func resourceService(serviceDef ServiceDefinition) *schema.Resource {
+		return resourceServiceCreate(ctx, d, meta, serviceDef)
 	}
 }
 
 // resourceRead satisfies the Terraform resource schema Read "interface"
 // while injecting the ServiceDefinition into the true Read functionality.
-func resourceRead(serviceDef ServiceDefinition) schema.ReadFunc {
-	return func(data *schema.ResourceData, i interface{}) error {
-		return resourceServiceRead(data, i, serviceDef, false)
+func resourceRead(serviceDef ServiceDefinition) schema.ReadContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		return resourceServiceRead(ctx, d, meta, serviceDef, false)
 	}
 }
 
 // resourceUpdate satisfies the Terraform resource schema Update "interface"
 // while injecting the ServiceDefinition into the true Update functionality.
-func resourceUpdate(serviceDef ServiceDefinition) schema.UpdateFunc {
-	return func(data *schema.ResourceData, i interface{}) error {
-		return resourceServiceUpdate(data, i, serviceDef)
+func resourceUpdate(serviceDef ServiceDefinition) schema.UpdateContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		return resourceServiceUpdate(ctx, d, meta, serviceDef)
 	}
 }
 
 // resourceDelete satisfies the Terraform resource schema Delete "interface"
 // while injecting the ServiceDefinition into the true Delete functionality.
-func resourceDelete(serviceDef ServiceDefinition) schema.DeleteFunc {
-	return func(data *schema.ResourceData, i interface{}) error {
-		return resourceServiceDelete(data, i, serviceDef)
+func resourceDelete(serviceDef ServiceDefinition) schema.DeleteContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+		return resourceServiceDelete(ctx, d, meta, serviceDef)
 	}
 }
 
 // resourceImport satisfies the Terraform resource schema Importer "interface"
 // while injecting the ServiceDefinition into the true Import functionality.
 func resourceImport(serviceDef ServiceDefinition) *schema.ResourceImporter {
 	return &schema.ResourceImporter{
-		State: func(d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
-			error := resourceServiceRead(d, m, serviceDef, true)
-			if error != nil {
-				return nil, error
+		StateContext: func(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+			err := diagToErr(resourceServiceRead(ctx, d, meta, serviceDef, true))
+			if err != nil {
+				return nil, err
 			}
 			return []*schema.ResourceData{d}, nil
 		},
 	}
 }
 
 // resourceServiceCreate provides service resource Create functionality.
-func resourceServiceCreate(d *schema.ResourceData, meta interface{}, serviceDef ServiceDefinition) error {
+func resourceServiceCreate(ctx context.Context, d *schema.ResourceData, meta interface{}, serviceDef ServiceDefinition) diag.Diagnostics {
 	if err := validateVCLs(d); err != nil {
-		return err
+		return diag.FromErr(err)
```

Comment:
Out of interest where are we using the new diagnostics and how does it affect what the end-user sees?

Reply:
This would be an example of the an error returned for the missing API key
```
$ terraform plan

Error: [Err] No API key for Fastly

  on <empty> line 0:
  (source code not available)
```
This is an example of the output for the replace warning
```
tls_private_key.key: Refreshing state... [id=e134c2b4c25235f0229aecb6e4ab7648cfb4655c]
fastly_tls_private_key.key: Refreshing state... [id=74NBnRVrCtX3gDrraqdvRE]
tls_self_signed_cert.cert: Refreshing state... [id=177197525387460537608140314827772347932]
fastly_service_v1.service: Refreshing state... [id=2u3G4BNkNQ4YtMK4hSz6kg]
fastly_tls_certificate.cert: Refreshing state... [id=5QNFZkSlBIo6zSuC9RWzPa]
fastly_tls_activation.ocdemo: Refreshing state... [id=32xc1xMqJB5ejLirCIM8Dw]

Warning: Fastly recommends that this certificate (5QNFZkSlBIo6zSuC9RWzPa) be replaced



Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```